miri-script/src/commands.rs

```diff
@@ -0,0 +1,652 @@
+        let sh = self.shell_with_miri_env()?;
+        let output = cmd!(sh, "cargo +{active_toolchain} run {cargo_extra_flags...} --manifest-path {manifest_path} -- miri setup --print-sysroot {target_prefix...} {target...}").read();
+        if output.is_err() {
+            // Run it again so the user can see the error.
+            // This time without printing a sysroot though.
+            let _ = cmd!(sh, "cargo +{active_toolchain} run {cargo_extra_flags...} --manifest-path {manifest_path} -- miri setup {target_prefix...} {target...}").run().with_context(|| "`cargo miri setup` failed")?;
+        }
+
+        Ok(output?)
+    }
+    /// Returns a toolchain override for use in command line (`+{toolchain_name}`).
+    fn get_toolchain_override(&self) -> OsString {
+        format!("+{}", &self.active_toolchain).into()
+    }
+    fn build_package(
+        // Path to Cargo.toml file of a package to build.
+        path: &OsStr,
+        toolchain: &OsStr,
+        extra_flags: &[String],
+        args: impl IntoIterator<Item = impl AsRef<OsStr>>,
+    ) -> Result<()> {
+        let sh = Shell::new()?;
+        cmd!(sh, "cargo {toolchain} build {extra_flags...} --manifest-path {path} {args...}")
+            .run()?;
+        Ok(())
+    }
+    fn shell_with_miri_env(&self) -> Result<Shell> {
+        let sh = shell_with_parent_env()?;
+        for (k, v) in &self.env {
+            sh.set_var(k, v);
+        }
+
+        Ok(sh)
+    }
+}
+
+impl MiriRunner<'_> {
+    fn bench(&self, benches: &[OsString]) -> Result<()> {
+        // The hyperfine to use
+        let hyperfine = std::env::var("HYPERFINE");
+        let hyperfine = hyperfine.as_deref().unwrap_or("hyperfine -w 1 -m 5 --shell=none");
+        let hyperfine = shell_words::split(hyperfine).unwrap();
+        let Some((program_name, args)) = hyperfine.split_first() else {
+            bail!("Expected HYPERFINE environment variable to be non-empty");
+        };
+        // Make sure we have an up-to-date Miri installed
+        Self::exec_inner(&Subcommands::Install { flags: vec![] }, false)?;
+        let benches_dir = path!(self.miri_dir / "bench-cargo-miri");
+        let benches = if benches.is_empty() {
+            std::fs::read_dir(&benches_dir)?
+                .filter_map(|path| {
+                    path.ok()
+                        .filter(|dir| dir.file_type().map(|t| t.is_dir()).unwrap_or(false))
+                        .map(|p| p.file_name())
+                })
+                .collect()
+        } else {
+            benches.to_owned()
+        };
+        let sh = shell_with_parent_env()?;
+        let toolchain = &self.active_toolchain;
+        // Run the requested benchmarks
+        for bench in benches {
+            let current_bench_dir = path!(benches_dir / bench / "Cargo.toml");
+            cmd!(
+                sh,
+                "{program_name} {args...} 'cargo +'{toolchain}' miri run --manifest-path '{current_bench_dir}''"
```

Comment:
What is going on with the `'` here?

Reply:
It is a special syntax from xshell; https://docs.rs/xshell/latest/xshell/macro.cmd.html It helps with escaping a bunch.